src/main/java/build/buildfarm/instance/AbstractServerInstance.java

```diff
@@ -177,78 +180,91 @@ private void environmentVariablesUniqueAndSortedPrecondition(
         ExecutePreconditionViolationType.ENVIRONMENT_VARIABLES_NOT_SORTED);
   }
 
-  private void validateAction(Action action) {
-    Digest commandDigest = action.getCommandDigest();
-    ImmutableList.Builder<Digest> inputDigests = new ImmutableList.Builder<>();
-    inputDigests.add(commandDigest);
-    Iterator<Directory> directoriesIterator =
-      createTreeIterator(action.getInputRootDigest(), "");
+  private void validateActionInputDirectory(
+      Directory directory,
+      Stack<Digest> path,
+      Set<Digest> visited,
+      ImmutableSet.Builder<Digest> inputDigests) {
```

Comment:
An empirical test of this indicates that ImmutableSet.Builder<> allows multiple adds to take place with the same value (a digest of the string "123") with no error and expected reduced behavior:

```
import build.buildfarm.common.Digests;
import com.google.common.collect.ImmutableSet;
import com.google.devtools.remoteexecution.v1test.Digest;
import com.google.protobuf.ByteString;

class test {
  public static void main(String[] args) {
    ImmutableSet.Builder<Digest> set = new ImmutableSet.Builder<>();

    set.add(Digests.computeDigest(ByteString.copyFromUtf8("123")));
    set.add(Digests.computeDigest(ByteString.copyFromUtf8("123")));

    System.out.println(set.build());
  }
}
```

```
$ test
[hash: "40bd001563085fc35165329ea1ff5c5ecbdbbeef"
size_bytes: 3
]
```